Result Page: Persist the column filters in the results table in the url

Demi-la opened the pull request:
Features: #765 

### Overview
This pull request fixed the issue of column filters not persisting after refreshing and also prevent re renders after clicking on a filter keyword.

### Changes made
The Previous implementation doesn't save the filter state in the URL or any persistent storage. When the page refreshes, the component reinitializes with an empty filter state. To fix this, I;

1. Saved the filter state in a URL parameters.
2. Initialize the filter state from the URL parameters when the component mounts.
3. Added a useEffect hook that updates the URL parameters whenever the tableFilters state changes. This ensures that the URL always reflects the current filter state.
4. Used  updateRawSearchParams from useRawSearchParams custom hook to prevent re-rendering.

https://github.com/user-attachments/assets/bbf259ca-d337-4205-b47a-c62079af14a6



### Testing
 Manual testing was performed to ensure the column filters persist and integrate seamlessly with existing functionalities.

Carla-Moz commented:
Will get to this tomorrow. Thanks for your patience!

Carla-Moz commented on src/components/CompareResults/ResultsTable.tsx:
```diff
@@ -103,9 +103,34 @@ export default function ResultsTable() {
+    }
+    return filters;
+  });
+
```
Adding a comment here before this useEffect would be helpful.

Carla-Moz commented on src/components/CompareResults/ResultsTable.tsx:
```diff
@@ -103,9 +103,34 @@ export default function ResultsTable() {
+      updateRawSearchParams(newSearchParams);
+    }
+  }, [tableFilters, searchParams, setSearchParams]);
 
```
Good job on this first commit! I suggest memoizing `URLSearchParams` (see [useMemo ](https://react.dev/reference/react/useMemo)for react) to avoid recreating it on every render. You can then add it the `useEffect`'s dependency array to make sure it's only recalculated when `searchParams` changes.

Demi-la commented:
@Carla-Moz  Thank you for the review and also providing the useMemo link.
I ran this command "npm run fix-all" but the problem is that there a lot  of files that i need to commit, i don't know if i should push all the files or just the file i worked on. Your suggestion would be highly appreciated.

Carla-Moz commented:
> @Carla-Moz Thank you for the review and also providing the useMemo link. I ran this command "npm run fix-all" but the problem is that there a lot of files that i need to commit, i don't know if i should push all the files or just the file i worked on. Your suggestion would be highly appreciated.

What files? Are you talking about the test snapshots?

Demi-la commented:
>What files? Are you talking about the test snapshots? Could you show me the list?
Not the snapshots, 
these are part of the files
  modified:   .circleci/config.yml
        modified:   .eslintrc.js
        modified:   .prettierrc.js
        modified:   @types/taskcluster-client-web.d.ts
        modified:   Deployment.md
        modified:   README.md
        modified:   babel.config.json
        modified:   jest.config.ts
        modified:   package-lock.json
        modified:   package.json
        modified:   public/index.html
        modified:   server.js
        modified:   src/__tests__/App.test.tsx
        modified:   src/__tests__/CompareResults/OverTimeResultsView.test.tsx
        modified:   src/__tests__/CompareResults/ResultsTable.test.tsx       
        modified:   src/__tests__/CompareResults/ResultsView.test.tsx
        modified:   src/__tests__/CompareResults/Retrigger.test.tsx
but i ran this command "git checkout -- ." to revert all modified files back to their last committed state.

Demi-la commented:
@Carla-Moz  I have update the PR by including comment before the useEffect and memoizing URLSearchParams

I realized that i change  19 pages even though i only made changes to one file. i think it's probably because i mistakenly merged instead of rebasing because i confirms that i only pushed one file. i added my terminal log  so that you can reference.

This was when i check my git status to be sure that i only have one file
![Screenshot (736)](https://github.com/user-attachments/assets/f76aa1ea-5570-4ae6-84f9-13d95e5fc09c)

But i could not push because the tip of my current branch was behind, i mistakenly merge to my branch instead of rebase.
![Screenshot (737)](https://github.com/user-attachments/assets/e0de6a79-f027-4b0e-bcd3-b5f7a6500be0)

After merging i still could not push because of the same issue so i had to rebase to my branch, then i was finally able to push the code.
![Screenshot (739)](https://github.com/user-attachments/assets/c8689b49-78cb-44dc-9ba5-ae5de70e2ec8)

Please kindly let me know if it affected the project or if there is a way i can fix it.
Thank you.

Carla-Moz commented:
@Demi-la Are you still having merge conflict issues?

Demi-la commented:
> @Demi-la Are you still having merge conflict issues?

No, I do not have merge conflict

Carla-Moz reviewed (commented):
Nice! Thanks for the changes

Demi-la commented:
> Nice! Thanks for he changes

Thank you for all your kind review and help.

Demi-la commented:
@Carla-Moz  Hello, i just noticed these conflicts, i'm confused because there are no conflicts in my local machine. 
The only conflict i had was in the ResultsTable file and i resolved it before pushing the code. but these conflicts are in different files.
![Screenshot (740)](https://github.com/user-attachments/assets/d9eefa30-b324-4a32-9a8e-a2b710642d06)
s

Carla-Moz commented:
> @Carla-Moz Hello, i just noticed these conflicts, i'm confused because there are no conflicts in my local machine. The only conflict i had was in the ResultsTable file and i resolved it before pushing the code. but these conflicts are in different files. ![Screenshot (740)](https://private-user-images.githubusercontent.com/113023546/379332648-d9eefa30-b324-4a32-9a8e-a2b710642d06.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3OTQzMTAsIm5iZiI6MTcyOTc5NDAxMCwicGF0aCI6Ii8xMTMwMjM1NDYvMzc5MzMyNjQ4LWQ5ZWVmYTMwLWIzMjQtNGEzMi05YThlLWEyYjcxMDY0MmQwNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyNFQxODIwMTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZjJmMGRlYTE5YjcwNTIyZTE3ZGYyNzk0OGIxZmE0ZTc4OTNhMjU4ZDU2YjhhNGVhNzQxOGYyNGE2MjVmMmEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.yqyUGHUZvzvJSL4jdHZ4HtkiJtCl9YBPhrKSKMSVvag) s

Are you sure you're local machine is up to date with the remote? 
# Git commands for keeping your branch up to date with the latest main
git fetch upstream
git rebase upstream/main
git push --force origin <local branch>

Demi-la commented:
> > @Carla-Moz Hello, i just noticed these conflicts, i'm confused because there are no conflicts in my local machine. The only conflict i had was in the ResultsTable file and i resolved it before pushing the code. but these conflicts are in different files. ![Screenshot (740)](https://private-user-images.githubusercontent.com/113023546/379332648-d9eefa30-b324-4a32-9a8e-a2b710642d06.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3OTQzMTAsIm5iZiI6MTcyOTc5NDAxMCwicGF0aCI6Ii8xMTMwMjM1NDYvMzc5MzMyNjQ4LWQ5ZWVmYTMwLWIzMjQtNGEzMi05YThlLWEyYjcxMDY0MmQwNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyNFQxODIwMTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZjJmMGRlYTE5YjcwNTIyZTE3ZGYyNzk0OGIxZmE0ZTc4OTNhMjU4ZDU2YjhhNGVhNzQxOGYyNGE2MjVmMmEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.yqyUGHUZvzvJSL4jdHZ4HtkiJtCl9YBPhrKSKMSVvag) s
> 
> Are you sure you're local machine is up to date with the remote? **Git commands for keeping your branch up to date with the latest main**
> 
> ```
> git fetch upstream
> git rebase upstream/main
> git push --force origin <local branch>
> ```

Yes it is. I still ran these command yesterday and there was no conflict locally. But let me do it again and push again maybe it will solve it

Demi-la commented:
@Carla-Moz  The merge conflict has been solved. I ran these commands 
git fetch upstream
git rebase upstream/main
git push --force origin <local branch>

Carla-Moz commented:
That's great! 

@Demi-la  Also, I'd like to let you know you've been selected to go on the next round! Congrats!
Please see the matrix channel for details https://chat.mozilla.org/#/room/#perfcompare-outreachy:mozilla.org

Demi-la commented:
> That's great!
> 
> @Demi-la Also, I'd like to let you know you've been selected to go on the next round! Congrats! Please see the matrix channel for details https://chat.mozilla.org/#/room/#perfcompare-outreachy:mozilla.org

Oh my Goodness!!!! Thank you for the good news
